src/AI/CapPhaseAnalyzer.java

```diff
@@ -0,0 +1,454 @@
+        for( CapStop stop : chain )
+          sb.append(String.format("    %s\n", stop));
+      }
+    }
+    return sb.toString();
+  }
+
+  /**
+   * If the input unit has a cap chain allocated, return it
+   * Else, if it's on the start of a cap chain in the un-allocated list, remove the chain from the list and allocate it
+   * @return The chosen chain, if any
+   */
+  public ArrayList<CapStop> getCapChain(Unit unit)
+  {
+    ArrayList<CapStop> chain = null;
+    XYCoord position = new XYCoord(unit);
+    // Add the unit to a cap chain, if possible
+    if( capChainsAllocated.containsKey(unit) )
+      chain = capChainsAllocated.get(unit);
+    else
+    {
+      ArrayList<ArrayList<CapStop>> chainList = capChains.get(position);
+      if( null != chainList && !chainList.isEmpty() )
+      {
+        chain = chainList.remove(0);
+        capChainsAllocated.put(unit, chain);
+        if( chainList.isEmpty() )
+          capChains.remove(position);
+      }
+    }
+
+    return chain;
+  }
+
+  public GameAction getCapAction(GameMap map, Unit unit)
+  {
+    XYCoord position = new XYCoord(unit);
+    ArrayList<CapStop> chain = getCapChain(unit);
+    if( null != chain )
+    {
+      if( unit.getCaptureProgress() > 0 )
+      {
+        return new CaptureLifecycle.CaptureAction(map, unit, Utils.findShortestPath(unit, position, map));
+      }
+
+      while (!chain.isEmpty())
+      {
+        final XYCoord coord = chain.get(0).coord;
+        // If we're already there or own it, throw the point out
+        if( position.equals(coord) || !unit.CO.isEnemy(map.getLocation(coord).getOwner()) )
+          chain.remove(0);
+        else
+          return findChainAction(map, unit, chain);
+      }
+      // Chain's done, if we ever get here
+      capChainsAllocated.remove(unit);
+    }
+
+    return null;
+  }
+
+  public GameAction findChainAction(GameMap gameMap, Unit unit, ArrayList<CapStop> chain)
+  {
+    XYCoord goal = chain.get(0).coord;
+
+    boolean includeOccupiedSpaces = false; // We assume that the area won't be cluttered
+    ArrayList<XYCoord> destinations = Utils.findPossibleDestinations(unit, gameMap, includeOccupiedSpaces);
+
+    // If we can get to our destination, go for it
+    if( destinations.contains(goal) )
+      return new CaptureLifecycle.CaptureAction(gameMap, unit, Utils.findShortestPath(unit, goal, gameMap));
+
+    Utils.sortLocationsByTravelTimeToCoord(unit, destinations, gameMap, goal);
+
+    for( XYCoord moveCoord : destinations )
+    {
+      Unit resident = gameMap.getLocation(moveCoord).getResident();
+      boolean spaceFree = null == resident;
+      if( !spaceFree )
+        continue; // I don't know how we got here, but bail
+
+      // Figure out how to get here.
+      GamePath movePath = Utils.findShortestPath(unit, moveCoord, gameMap);
+
+      return new WaitLifecycle.WaitAction(unit, movePath);
+    }
+    return null;
+  }
+
+  public CapPhaseAnalyzer(GameMap map, Army viewer)
```

Comment:
Y'know, most people would put the class constructor at the top of the class, below the class variables, but... you do you, I guess.

Reply:
My reasoning was basically "put the basic API first, and the constructor goes later 'cause it's super long and involved"
Maybe I should make most of that a named function and make a smaller constructor that calls it.

Reply:
Might or might not help with clarity. Maybe not worthwhile if the new functions wouldn't be called from anywhere else.

Reply:
Yeah. It'd only be called from there.
That's probably why I didn't do it.

It might be worth breaking that function down a bit, but I want readable function signatures and that's hard to do with so much reused state.
Could probably package things into structs to hold that specific data, but I'm not sure that actually makes things clearer.